src/mbgl/renderer/sources/render_geojson_source.cpp

```diff
@@ -38,8 +38,9 @@ void RenderGeoJSONSource::update(Immutable<style::Source::Impl> baseImpl_,
         return;
     }
 
-    if (data_ != data) {
+    if (data_ != data || data_->needsUpdate) {
```

Comment:
Given that we're already checking `needsUpdate`, we can remove the pointer comparison altogether.